### DISCUSSION TITLE
API connecction to tactic

### DISCUSSION THREAD

**Discussion by DittoLabsNg** [listyque/TACTIC-Handler#44] [General]
Please what part of the code connects to the tactic server, does authentication grabs data from tactic?

**Comment by listyque** [comment]
Hi!
Basic auth to tactic api-server would look like this:
```
from client.tactic_client_lib.tactic_server_stub import TacticServerStub

host = 'localhost:9123'
project = 'sthpw'
login = 'LOGIN'
password = 'PASSWORD'

server = TacticServerStub(protocol='xmlrpc', setup=False)

server.set_server(host)
server.set_project(project)

ticket = server.get_ticket(login, password)
server.set_ticket(ticket)
```

After that you can use server.
Ticket can be generated once, and then reused until it expires. And after it expires regenerate it again.

**Reply by DittoLabsNg** [reply]
Thanks a lot

**Reply by listyque** [reply]
Example above shows how to use "pure" Tactic Client Api.
If you want to use Tactic-Handler api wrap this is how pre-setup Tactic-Handler:
```
import sys
INSTALLED_PATH = '/home/tactic_handler'  #folder where you unpacked T-H
SERVER_NAME = 'localhost:9123'

def preconfigure_tactic_handler(repo_path=None, login=None, server_url=None):

    if INSTALLED_PATH not in sys.path:
        sys.path.append(INSTALLED_PATH)

    import thlib.environment as thenv
    thenv.env_mode.set_current_path(INSTALLED_PATH)
    thenv.env_mode.set_mode('maya')

    if login:
        print('setting login', login)
        thenv.env_server.set_server(
            server_name=server_url,
        )
        thenv.env_server.set_timeout(
            timeout=180,
        )
        thenv.env_server.set_user(
            user_name=login,
        )

        thenv.env_server.set_site(
            site_name='',
            enabled=False,
        )

        thenv.env_server.set_proxy(
            proxy_login='',
            proxy_pass='',
            proxy_server='',
            enabled=False,
        )

        kwargs = thenv.tc().generate_new_ticket(login)
        thenv.tc().server_auth(**kwargs)

        if thenv.env_server.get_ticket():
            thenv.env_server.set_ticket(thenv.env_server.get_ticket())

        thenv.env_server.save_server_presets_defaults()
        thenv.env_server.save_defaults()

    if repo_path:
        thenv.env_tactic.get_base_dirs()

        print('setting repo', repo_path)
        base_dir = [u'{0}'.format(repo_path), u'General', [128, 128, 128], u'base', True]
        thenv.env_tactic.set_base_dir('base', base_dir)
        thenv.env_tactic.save_base_dirs()


login = 'admin'
repo_path = '/home/assets'
preconfigure_tactic_handler(repo_path, login, SERVER_NAME)
```
Note:
You can paste this code into a .py file, save it in same folder where you unpacked T-H, and execute it
This will ask you to enter login and password to generate ticket.

**Reply by DittoLabsNg** [reply]
Thanks a lot.

I've been trying to run the client API code for checkin but it keeps saying "cannot import module common".
When it runs the upload_file method in simple_checkin() it tries to import UploadMultipart class from common and throws an error that the common module is not available.
I really don't know the cause of this error or how to go about it

**Comment by DittoLabsNg** [comment]
Based on my last comment, my main question is, what part of the code base handles checkin? I need this to be able to know why it keeps throwing an error

**Reply by listyque** [reply]
Looks like you have issues with modules import.
How you are trying to execute client code?

**Reply by DittoLabsNg** [reply]
Okay so I know this is not really professional or the best way to implement this but, I created a Django app with Django REST Framework as a web API to get data and files from my tactic app . 
My main aim is to communicate with this API from Toon Boom Harmony.

I copied the tactic client directory just like you said and put it in the Django src directory then I added the client API codes(from Tactic documentation) to the Django views that will get and post data.

The code you sent here earlier on connecting to the API worked and I could return the ticket to the API as response but, when I tried the code for checkin on the tactic documentation, it keeps throwing that error ---> No module named 'common'.

Please what exactly am I doing wrong?

**Reply by listyque** [reply]
If you're not planning to use Tactic-Handler, then you can address your questions directly to TACTIC developers, at: [http://forum.southpawtech.com](url)

The thlib/side/ folder holds completely separate code, whis is not maintained by me.

**Comment by DittoLabsNg** [comment]
Hi based on my last comment, I already solved the issue by adding path "client/tactic_client_lib" to the sys.path and it solved the problem.
Thanks a lot.

After this it showed another error  'str' object has no attribute 'decode'.
I think this comes from the execute() method in the UploadMultipart class, it tries to decode a string.
